storage/engine: unify code to commit write-only and read-write batches

petermattis opened the pull request:
Allow grouping read-write batches into batch groups. If the leader is a
read-write batch, we create a temporary write-only batch to perform the
grouping. This addresses the performance hit seen in #16951, but also
highlights that the expected boost from that PR is likely to be small.

petermattis committed:
storage/engine: unify code to commit write-only and read-write batches

Allow grouping read-write batches into batch groups. If the leader is a
read-write batch, we create a temporary write-only batch to perform the
grouping. This addresses the performance hit seen in #16951, but also
highlights that the expected boost from that PR is likely to be small.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16976)
<!-- Reviewable:end -->

petermattis committed:
storage/engine: lazily initialize the RocksDB batch

Lazily initialize the RocksDB batch and add support for avoiding Cgo
when applying a batch repr to a write-only batch. For the write-only
batches uses in `Replica.applyRaftCommand()`, this avoids creating the
RocksDB batch completely, saving 4 Cgo calls per Raft command if the
batch is committed as part of a group (batch creation, batch
destruction, applying the batch repr and retrieving the batch
repr). Testing on a production cluster with a write-only workload showed
a reduction in Cgo calls from ~800k/sec (across 6 nodes) to ~200k/sec.

petermattis commented:
Added a 2nd commit which reduces the number of Cgo calls during write operations. Visual inspection of throughput graphs show a modest benefit. I have to run longer tests to have more precise numbers.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16976#-:-Komo0-HQ3VMVsFX-ocO:bnfp4nl)*
<!-- Sent from Reviewable.io -->
